editor/animation_track_editor.cpp

```diff
@@ -7709,3 +7859,1111 @@ AnimationTrackKeyEditEditor::AnimationTrackKeyEditEditor(Ref<Animation> p_animat
+	menu->add_shortcut(ED_SHORTCUT("animation_marker_edit/toggle_marker_names", TTR("Show All Marker Names"), Key::M), MENU_KEY_TOGGLE_MARKER_NAMES);
+
+	marker_insert_confirm = memnew(ConfirmationDialog);
+	marker_insert_confirm->set_title(TTR("Insert Marker"));
+	marker_insert_confirm->set_hide_on_ok(false);
+	marker_insert_confirm->connect("confirmed", callable_mp(this, &AnimationMarkerEdit::_confirm_insert_marker));
+	add_child(marker_insert_confirm);
+	VBoxContainer *marker_insert_vbox = memnew(VBoxContainer);
+	marker_insert_vbox->set_anchors_and_offsets_preset(Control::LayoutPreset::PRESET_FULL_RECT);
+	marker_insert_confirm->add_child(marker_insert_vbox);
+	marker_insert_new_name = memnew(LineEdit);
+	marker_insert_new_name->connect("text_changed", callable_mp(this, &AnimationMarkerEdit::_marker_insert_new_name_changed));
+	marker_insert_confirm->register_text_enter(marker_insert_new_name);
+	marker_insert_vbox->add_child(_create_hbox_labeled_control(TTR("Marker Name"), marker_insert_new_name));
+	marker_insert_color = memnew(ColorPickerButton);
+	marker_insert_color->set_edit_alpha(false);
+	marker_insert_color->get_popup()->connect("about_to_popup", callable_mp(EditorNode::get_singleton(), &EditorNode::setup_color_picker).bind(marker_insert_color->get_picker()));
+	marker_insert_vbox->add_child(_create_hbox_labeled_control(TTR("Marker Color"), marker_insert_color));
+	marker_insert_error_dialog = memnew(AcceptDialog);
+	marker_insert_error_dialog->set_ok_button_text(TTR("Close"));
+	marker_insert_error_dialog->set_title(TTR("Error!"));
+	marker_insert_confirm->add_child(marker_insert_error_dialog);
+}
+
+AnimationMarkerEdit::~AnimationMarkerEdit() {
+}
+
+float AnimationMarkerKeyEdit::get_time() const {
+	return animation->get_marker_time(marker_name);
+}
+
+void AnimationMarkerKeyEdit::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("_hide_script_from_inspector"), &AnimationMarkerKeyEdit::_hide_script_from_inspector);
+	ClassDB::bind_method(D_METHOD("_hide_metadata_from_inspector"), &AnimationMarkerKeyEdit::_hide_metadata_from_inspector);
+	ClassDB::bind_method(D_METHOD("_dont_undo_redo"), &AnimationMarkerKeyEdit::_dont_undo_redo);
+	ClassDB::bind_method(D_METHOD("_is_read_only"), &AnimationMarkerKeyEdit::_is_read_only);
+	ClassDB::bind_method(D_METHOD("_set_marker_name"), &AnimationMarkerKeyEdit::_set_marker_name);
+}
+
+void AnimationMarkerKeyEdit::_set_marker_name(const StringName &p_name) {
+	marker_name = p_name;
+}
+
+bool AnimationMarkerKeyEdit::_set(const StringName &p_name, const Variant &p_value) {
+	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+	if (p_name == "name") {
+		const StringName &name = p_value;
+		marker_name_cache = name;
+		if (name != marker_name) {
+			if (!name) {
+				WARN_PRINT_ED(TTR("Empty marker names are not allowed."));
+			} else if (name != marker_name && animation->has_marker(name)) {
+				WARN_PRINT_ED(vformat(TTR("Marker %s already exists."), name));
```

Comment:
```suggestion
			if (!name) {
				ERR_FAIL_V_MSG(false, TTR("Empty marker names are not allowed."));
			} else if (name != marker_name && animation->has_marker(name)) {
				ERR_FAIL_V_MSG(false, vformat(TTR("Marker %s already exists."), name));
```

If an action is not allowed, it should be an error instead of a warning. Also, if there is an error, the input should be aborted.

Reply:
Fixed. Note that for better UX the illegal input is still shown in the inspector, because otherwise the text cursor would be moved along the previous legal input. For example:
- The current input is `test2`, marker `test` already exists, and the text cursor is behind "2".
- Entering backspace would result in an illegal input (`test`), so the text is still left unchanged as `test2`. However, the text cursor would be moved in front of 2 as a result of the backspace.

This would be annoying if the user is trying to enter a legal marker name but their edit sequence happens to trip over illegal inputs.

Reply:
Can you add an error dialog to the inspector plugin? Although it may be necessary to revert to the previous name at that time.

Or, a simpler implementation would be to make the name readonly on the inspector, and add Rename Marker option to the right-click menu. Since the inspector is primarily for things that may change frequently, such as color and time, I see it is fine with that.

Reply:
Would this work?


https://github.com/godotengine/godot/assets/56677134/d8e7f539-c91b-412b-9723-0558e3a30f2a

Reply:
Having a button there is not recommended because it is not consistent with other interfaces. I mean to add "Rename Marker" and a name-only change option to the right-click menu if there is to be a separate dialog like:

![image](https://github.com/godotengine/godot/assets/61938263/57dbc366-1d59-4601-bd4d-6419333102f4)

Reply:
> Then, the "Name" property must be Readonly (only on the inspector) or displayed as a Label instead of a LineEdit in the inspector. If you write in the Class reference that "the marker name must be changed from TrackEditor on the editor because it does not allow duplicate names", the Tooltip will inform the user at hover time.

The property is read-only in the inspector. Though, I don't think there's a way to add class reference without adding an entry to `doc/classes` (which would also cause doctool to pick it up and show it in the documentation website), for the otherwise-internal class.

Alternatively I've made it so the marker name is shown in the inspector plugin instead, so I can insert a tooltip with `set_tooltip_text`. This does look different from the tooltip that would be shown via `AnimationMarkerKeyEdit`'s `get_property_list`, however. Would this work?

![image](https://github.com/godotengine/godot/assets/56677134/2e5eb44a-496b-4ce9-a6d4-f87913097b0b)

Reply:
If the property is correctly assigned to the inspector editor, you should get a reference.
![image](https://github.com/godotengine/godot/assets/61938263/03809398-380d-4bbb-ab28-9d022f57af72)

Reply:
I meant that it's impossible (to my knowledge at least) to add description to the property's tooltip without adding a file to `doc/classes` (e.g., `AnimationMarkerKeyEdit.xml`). Doing so might not be a good idea because this file would also show up on the official documentation site as a side effect.

Reply:
Or, in such a case, what we have already done elsewhere is to add a warning. See `InspectorDock::set_info()`.

![image](https://github.com/godotengine/godot/assets/61938263/f4c3a648-85d8-48d0-a5c0-c0c1f87c4c50)

Reply:
How about now?

https://github.com/godotengine/godot/assets/56677134/c1e431b1-f312-4304-a04e-69d2e9322e01

Reply:
Previous Name is not required in popup, you can refer to rename dialog with other resources.

Reply:
Also, for the yellow text in the Warning, "Marker name is read-only in the EditorInspector." would be better. Not everyone is trying to change their name there, but they are being warned.

Reply:
Can you match Animation resource renaming dialog?

![image](https://github.com/godotengine/godot/assets/61938263/0cfc02e1-6db2-420a-b155-545ef9d0e108)

Also, the message should be "EditorInspector" or "inspector", not "Editor Inspector". If you prefer a shorter, then "inspector" may be a better choice.

Reply:
Done.

![image](https://github.com/godotengine/godot/assets/56677134/7fc999f7-0445-47f6-8739-fb7e4b460e90)

Reply:
Is it possible to replicate the behavior of the Animation resource renaming dialog, not just the message?

Reply:
Like this?

https://github.com/godotengine/godot/assets/56677134/50ace68e-bd1e-449c-b25b-a338bca90061

Reply:
It's fine.